File: backend/app/models/device.py
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, DateTime, ForeignKey, Enum as SAEnum, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base
import enum


class DeviceStatus(str, enum.Enum):
    locked = "locked"
    unlocked = "unlocked"
    offline = "offline"
    tamper = "tamper"


class DeviceType(Base):
    __tablename__ = "device_types"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    code: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    capabilities: Mapped[dict | None] = mapped_column(JSONB)
    sensor_schema: Mapped[dict | None] = mapped_column(JSONB)

    devices = relationship("Device", back_populates="device_type")


class Device(Base):
    __tablename__ = "devices"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    device_type_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("device_types.id"), nullable=False)
    owner_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    mac_address: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    firmware_version: Mapped[str | None] = mapped_column(String(20))
    status: Mapped[DeviceStatus] = mapped_column(
        SAEnum(DeviceStatus, name="device_status", create_type=False),
        nullable=False,
        default=DeviceStatus.offline,
    )
    battery_level: Mapped[int | None] = mapped_column(Integer)
    location: Mapped[str | None] = mapped_column(String(100))
    paired_at: Mapped[datetime | None] = mapped_column(DateTime)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

    device_type = relationship("DeviceType", back_populates="devices")
    owner = relationship("User", back_populates="devices", foreign_keys=[owner_id])
    access_cards = relationship("AccessCard", back_populates="device")
    access_logs = relationship("AccessLog", back_populates="device")

File: smart_lock_app/lib/core/models/device.dart
enum DeviceStatus { locked, unlocked, offline, tamper, unknown }

class SmartDevice {
  final String id;
  final String name;
  final String? location;
  final DeviceStatus status;
  final int? batteryLevel;
  final String macAddress;
  final String? firmwareVersion;
  final String? ownerName;
  final DateTime? updatedAt;

  SmartDevice({
    required this.id,
    required this.name,
    this.location,
    required this.status,
    this.batteryLevel,
    required this.macAddress,
    this.firmwareVersion,
    this.ownerName,
    this.updatedAt,
  });

  factory SmartDevice.fromJson(Map<String, dynamic> json) {
    return SmartDevice(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? 'Unknown',
      location: json['location'],
      status: _parseStatus(json['status']),
      batteryLevel: json['battery_level'],
      macAddress: json['mac_address'] ?? '',
      firmwareVersion: json['firmware_version'],
      ownerName: json['owner_name'],
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'])
          : null,
    );
  }

  static DeviceStatus _parseStatus(dynamic value) {
    if (value == null) return DeviceStatus.unknown;
    switch (value.toString().toLowerCase()) {
      case 'locked':
        return DeviceStatus.locked;
      case 'unlocked':
        return DeviceStatus.unlocked;
      case 'offline':
        return DeviceStatus.offline;
      case 'tamper':
        return DeviceStatus.tamper;
      default:
        return DeviceStatus.unknown;
    }
  }

  bool get isLocked => status == DeviceStatus.locked;
  bool get isOnline => status != DeviceStatus.offline && status != DeviceStatus.unknown;
  bool get isTamper => status == DeviceStatus.tamper;

  String get statusLabel {
    switch (status) {
      case DeviceStatus.locked:
        return 'Đã khóa';
      case DeviceStatus.unlocked:
        return 'Đã mở';
      case DeviceStatus.offline:
        return 'Offline';
      case DeviceStatus.tamper:
        return 'Báo động';
      default:
        return 'Không rõ';
    }
  }

  SmartDevice copyWith({
    DeviceStatus? status,
    int? batteryLevel,
  }) {
    return SmartDevice(
      id: id,
      name: name,
      location: location,
      status: status ?? this.status,
      batteryLevel: batteryLevel ?? this.batteryLevel,
      macAddress: macAddress,
      firmwareVersion: firmwareVersion,
      ownerName: ownerName,
      updatedAt: updatedAt,
    );
  }
}
